runtime/onert/api/src/nnfw_api_internal.cc

```diff
@@ -1312,15 +1312,39 @@ NNFW_STATUS nnfw_session::train_set_expected(uint32_t index, const void *expecte
     return NNFW_STATUS_INVALID_STATE;
   }
 
-  // Check index is valid: [0, getExpectedSize())
+  if (index >= getOutputSize())
+  {
+    std::cerr << "Error during nnfw_session::train_set_expected : index is out of range"
+              << std::endl;
+    return NNFW_STATUS_ERROR;
+  }
 
-  // Maybe it cannot use general execute->setOutput()
-  // because it needs to set training output (maybe loss)
-  (void)index;
-  (void)expected_tensorinfo;
+  try
+  {
+    auto output_ind = onert::ir::IOIndex(index);
+    auto size = _execution->getOutputTotalSize(output_ind);
+    if (expected_tensorinfo && getBufSize(expected_tensorinfo) != size)
+    {
+      std::cerr << "Error during nnfw_session::train_set_expected : invalid tensorinfo"
+                << std::endl;
+      return NNFW_STATUS_ERROR;
+    }
 
-  // NYI
-  return NNFW_STATUS_ERROR;
+    // NOTE Find the loss input index
+    // Input is added as many as the number of outputs.
+    // The loss index is calculated from the value obtained by subtracting the
```

Comment:
Exactly, it does not calculate the loss index, but the index of _expected_, which is the input of _loss_.